src/lib.rs

```diff
@@ -9,30 +9,60 @@
 //! current tests are not exhaustive.
 //!
 //! # Example
-//!
-//! ```rust,no_run
-//! # fn try_main() -> Result<(), Box<dyn std::error::Error>> {
-//! // Read events from `events_in.hepmc2` and write them to `events_out.hepmc2`
-//! use hepmc2::{Reader, Writer};
-//!
-//! use std::io::BufReader;
-//! use std::fs::File;
-//!
-//! let input = BufReader::new(File::open("events_in.hepmc2")?);
-//! let in_events = Reader::from(input);
-//!
-//! let output = File::create("events_out.hepmc2")?;
-//! let mut writer = Writer::try_from(output)?;
-//!
-//! for event in in_events {
-//!    let event = event?;
-//!    println!("Current cross section: {}",  event.xs);
-//!    writer.write(&event)?
-//! }
-//! writer.finish()?;
-//! # Ok(())
-//! # }
-//! ```
+#![cfg_attr(
+    feature = "sync",
+    doc = r##"
+```rust,no_run
+// Read events from `events_in.hepmc2` and write them to `events_out.hepmc2`
+use hepmc2::{Reader, Writer};
+
+use std::io::BufReader;
+use std::fs::File;
+
+let input = BufReader::new(File::open("events_in.hepmc2")?);
+let in_events = Reader::from(input);
+
+let output = File::create("events_out.hepmc2")?;
+let mut writer = Writer::try_from(output)?;
+
+for event in in_events {
+   let event = event?;
+   println!("Current cross section: {}",  event.xs);
+   writer.write(&event)?
+}
+writer.finish()?;
+# Ok::<(), Box<dyn std::error::Error>>(())
+```"##
+)]
+#![cfg_attr(
+    feature = "tokio",
+    doc = r##"
+```rust,no_run
+# async fn try_main() -> Result<(), Box<dyn std::error::Error>> {
+// Read events from `events_in.hepmc2` and write them to `events_out.hepmc2`
+use hepmc2::{Reader, Writer};
+
+use tokio::io::BufReader;
+use tokio::fs::File;
+
+let input = BufReader::new(File::open("events_in.hepmc2").await?);
+let mut in_events = Reader::from(input);
+
+let output = File::create("events_out.hepmc2").await?;
+let mut writer = Writer::try_from(output).await?;
+
+while let Some(event) = in_events.next().await {
+   let event = event?;
+   println!("Current cross section: {}",  event.xs);
+   writer.write(&event).await?
+}
+writer.finish().await?;
+# Ok(())
+# }
+# tokio_test::block_on(async {try_main().await.unwrap()})
+```"##
+)]
+
```

Comment:
I think it would make sense to have _both_ the sync and the async example code in the documentation, regardless of which features are enabled. We can then put a `#[cfg]` inside the examples to make the actual compilation depend on the features.

At the moment, the doc here and the readme are the same - the readme is in fact generated with `cargo readme`. I would like to keep it that way, i.e. we would copy the new text in the readme over to the doc here.

Reply:
Done. Found a way to do this with `cfg_attr` outside of the examples, which I think I prefer.

I'm afraid though the readme no longer matches `cargo-readme` - it can't handle doc tests gated behind features: https://github.com/webern/cargo-readme/issues/54